### PR TITLE
Use more state machine like state management in progressService

### PR DESCRIPTION
**Pull request by mjbvz**
**Problem**
In progress service, a `ProgressState` object tracks state. How this object works is difficult to reason about. `ProgressState` is an interface consisting only of optional fields, however only a subset of fields are (or should be) active at a given time. Code that interacts with the `ProgressState` object is difficult to understand and not easy to safely update

When enabling strict null checks, the existing `ProgressState` is also a pain because the type system does not know which subset of fields will be active together. This can result in lots of extra type checking

**Fix**
I've been experimenting with writing state more explicitly in our codebase and have already moved the parameter hints and code action models to use a more state machine like approach to state. This has two main features:

- Encapsulating all state in a single state object. The progress service already did this.
- Encoding valid states as a set of explicit state types

For progress service, the encoding breaks the `ProgressState` object up into five different states that the progress service transitions between. Refactoring code to make state more explicit has some key benefits:

- Better type system support, especially with strict null checks
- Catching programming errors, such as invalid transitions
- The code becomes more readable (at least once you get used to the pattern)
- Better comprehension of the program logic since state changes are more explicit

This PR demonstrates the pattern and may be of interest for other places in the code base

### CODE REVIEW

**Review by bpasero — APPROVED** [review]
LGTM. Of course a large change like this can introduce new bugs that are not easy to spot through a quick review, so maybe we could add some tests to verify the logic.

File: src/vs/workbench/services/progress/browser/progressService.ts [hunk 1]
```diff
@@ -10,16 +10,39 @@ import { IViewletService } from 'vs/workbench/services/viewlet/browser/viewlet';
...
-interface ProgressState {
-	infinite?: boolean;
-	total?: number;
-	worked?: number;
-	done?: boolean;
-	whilePromise?: Promise<any>;
-	whileStart?: number;
-	whileDelay?: number;
+
+const NoneProgressState = new class { readonly type = 'none'; };
```
Comment (alexandrudima) [inline src/vs/workbench/services/progress/browser/progressService.ts]: @mjbvz I always use const enums for such cases, I prefer the lower memory overhead of using integers compared to strings and that I can find all places where the type is checked with a simple search of the enum name. Plus, const enums get inlined (as ints) in the JS output even across different modules. An enum has the added benefit of immediately communicating that this is a closed set of types. i.e.

```ts
const enum ProgressType {
 None,
 Done,
 Infinite,
 While,
 Work
}

const NoneProgressState = new class { readonly type = ProgressType.None; };

// .. etc
```

**Review by alexdima — APPROVED** [review]
LGTM, maybe use a const enum instead of strings.

File: src/vs/workbench/services/progress/browser/progressService.ts (hunk 1 above)
Comment (mjbvz) [reply src/vs/workbench/services/progress/browser/progressService.ts]: That makes sense. I've updated this PR to use a const enum and move all the relevant types into a namespace

**Pull request merged**